### DISCUSSION TITLE
How to process multiple DDS cubemaps?

### DISCUSSION THREAD

**Discussion by IgeNiaI** [ImageMagick/ImageMagick#1988] [General] [answered]
Hello,

I have lots of DDS cubemaps that I want to modify, but I can't find any proper documentation on how to achieve what I want.
The image consists of 6 layers (+x, -x, +y, -y, +z, -z)
I want to rotate +x counter-clockwise, rotate -x clockwise, rotate +y by 180 degrees and then swap +y and -z, swap -y and +z.
What would be the ImageMagick command to do that?

Thanks in advance.

**Comment by dlemstra** [comment]
Could you attach the image to this post and explain in more detail (visually?) what you want to achieve?

**Reply by IgeNiaI** [reply]
I am making a videogame and I would like to use dds cubemaps as a skybox. I have over 100 images, but all of them are rotated incorrectly.

Wrong:
![Wrong](https://user-images.githubusercontent.com/9607628/81085360-204f7080-8ec5-11ea-8df7-cad8d1ec1874.png)
Correct:
![Correct](https://user-images.githubusercontent.com/9607628/81085356-1f1e4380-8ec5-11ea-9011-3432425fa562.png)

The actions listed above fix it, but doing it manually for over 100 images takes time. I actually wrote a script for GIMP that automates it, but still would like to know how to do it in ImageMagick.

**Reply by fmw42** [reply]
What do you mean by rotate +x clockwise? By how much? 90 deg?

**Reply by IgeNiaI** [reply]
Yes, it's 90 degrees. +x is the cubeface at the positive side along x axis.

**Comment by fmw42** [comment]
We need your input images for each cube face. Also what is your version of Imagmagick and platform/OS, as syntax may vary.

**Reply by IgeNiaI** [reply]
Here is the original and fixed image used in my example. I am using latest ImageMagick on Windows 10.

Original:
https://drive.google.com/file/d/1egHCXiiiAN4m-w4V9aZHc2t-EVz-jqCD/view?usp=sharing
Fixed:
https://drive.google.com/file/d/1ecPLNz5qM4lFr1nJXbDYH6wUT3qoc2zS/view?usp=sharing

**Reply by fmw42** [reply]
This is untested as I am not on Windows. Add the suffixes and proper names of each image in my command. Note that imagemagick indices start at 0.

```
convert plusX minusX plusY minusY plusZ minusZ ^
( -clone 0 -rotate -90 ) ^
( -clone 1 -rotate +90 ) ^
( -clone 2 -rotate 180 ) ^
-swap 8,5 ^
-swap 4,3 ^
-delete 0,1,2 ^
+write 0 plusX ^
+write 1 minusX ^
+write 2 plusY ^
+write 3 minusY ^
+write 4 plusZ ^
+write 5 minusZ ^
null:


```

**Reply by IgeNiaI** [reply]
I don't have 6 images for each face, I have one .dds cubemap image that I need to modify.

**Reply by fmw42** [reply]
But the DDS has layers that you can access.

`convert DDS:12-00_original.dds x.png`

Produces 6 images. Please label them as your +x, -x, +y, -y, +z, -z, so that we know the order.

![x-0](https://user-images.githubusercontent.com/13458218/81437383-eded7f80-911f-11ea-9db2-c58c63f57e4d.png)

![x-1](https://user-images.githubusercontent.com/13458218/81437387-f04fd980-911f-11ea-91a6-18ae89e403b1.png)

![x-2](https://user-images.githubusercontent.com/13458218/81437402-f5148d80-911f-11ea-95cd-9d720046c499.png)

![x-3](https://user-images.githubusercontent.com/13458218/81437411-f940ab00-911f-11ea-9b07-1f644dca4176.png)

![x-4](https://user-images.githubusercontent.com/13458218/81437415-fc3b9b80-911f-11ea-905e-e469ba71e485.png)

![x-5](https://user-images.githubusercontent.com/13458218/81437423-0067b900-9120-11ea-92df-3ce55465e490.png)

**Reply by IgeNiaI** [reply]
The original order is what you posted, aka +x, -x, +y, -y, +z, -z
In the final result these images should become +x, -x, +z, -z, -y, +y
Then the +x should rotate 90 degrees counterclockwise,
the -y should rotate 90 degrees clockwise and
the +y should rotate 180 degrees

**Comment by fmw42** [comment]
I want you to identify which of your labels goes with which of the 6 images by name, so I can order them appropriates.

My images are x-0.png, x-1.png, ... x-5.png.  Which one is +x, -x, +y, -y, +z, -z?

**Reply by IgeNiaI** [reply]
I told that in my comment.
Anyway,
x-0 is +x
x-1 is -x
x-2 is +y
x-3 is -y
x-4 is +z
x-5 is -z

**Reply by fmw42** [reply]
Thanks. Sorry, I overlooked that.

**Comment by fmw42** [comment]
What tool created your original DDS file? Is there a description of how the 6 images are formatted in the DDS file?  I do not normally works with DDS files. I do not know if ImageMagick can write the DDS in a way that it is compatible with your cube map viewer. I can read the DDS, convert to 6 PNG images, do the processing as in the code above to rotate and swap images, but I do not know how to take those 6 modified images and re-create a DDS file that is in cube map format.

**Reply by IgeNiaI** [reply]
I don't know. These cubemaps are extracted from a 13 years old game.
Based on Google, imagemagick can only create mipmap dds files, that's why I decided to ask this question here.

**Reply by fmw42** [reply]
What was the name of the game?

**Comment by dlemstra** [comment]
With the define `-define dds:mipmaps=fromlist` you can use the list of images to be use as the mipmaps of the output image.

**Reply by fmw42** [reply]
Dirk, those images are not mip-maps. The are for cube mapping and need to be in a specific format and likely have certain meta data consistent with the computer game from which they came so that they can be displayed 360 degrees in both directions with a compatible viewer. See https://en.wikipedia.org/wiki/Cube_mapping

**Reply by IgeNiaI** [reply]
Sorry, forgot to post the game name. It's S.T.A.L.K.E.R. series. These cubemaps are from a mod called Call of Chernobyl, but the skybox approach didn't change since Shadow of Chernobyl.
Unreal Engine 4 also uses DDS cubemaps, but I have to resave it using GIMP first, even if I don't want to modify anything.

**Reply by fmw42** [reply]
How do you create a compatible DDS file by modifying your image in GIMP?  Does GIMP know how to put the 6 images into a DDS file that is compatible with your other tools?

**Reply by IgeNiaI** [reply]
In GIMP just save it as a dds cubemap, it will then work in UE4. But I faced an issue there. Latest GIMP couldn't save cubemaps for some reason, so I had to install an older version (2.8.22) and a dds plugin for it.

**Reply by fmw42** [reply]
I cannot speak for the ImageMagick developers, but they would need some documentation on the DDS file format for cube maps that are compatible with your software, if they were to implement anything. What was the DDS plugin you used?

I tried to open your original DDS file in GIMP 2.10 on my Mac and it would not recognize the file format.

**Reply by IgeNiaI** [reply]
This is the plugin
https://code.google.com/archive/p/gimp-dds/

It was later integrated into GIMP, but it didn't work properly for me.

**Reply by fmw42** [reply]
I did find this http://wiki.simigon.com/wiki/index.php?title=Dds_cubemaps

**Reply by IgeNiaI** [reply]
Also the Nvidia tool for cubemaps
https://developer.nvidia.com/nvidia-texture-tools-exporter

**Reply by fmw42** [reply]
Your gimp-dds plug in appears to have code in C. I do not know if the IM developers can work from that or not. I tried to write your 6 original layers that I extracted back to a new DDS file, but it only shows one image and not all 6.

**Reply by IgeNiaI** [reply]
You need to choose to export as a cubemap.

**Reply by fmw42** [reply]
I could not even read the file in GIMP. 

The IM developers will have to comment further on this. IM can read your DDS file and export the 6 layer images. My script can reprocess them as you like. The missing piece is writing a new DDS file compatible with the DDS cube map format. Seems like ImageMagick is not writing the 6 files as layers in the DDS image.

**Reply by fmw42** [reply]
Have you tried any of the software at http://wiki.polycount.com/wiki/Cube_map

**Reply by IgeNiaI** [reply]
Writing a new dds cubemap is where I was stuck. I made a script for GIMP that does all the work except for saving the images, so I don't really need IM here.

**Reply by fmw42** [reply]
Well, that is up to the IM developers. But I did find this documenation. https://docs.microsoft.com/en-us/windows/win32/direct3ddds/dx-graphics-dds-pguide
http://doc.51windows.net/directx9_sdk/graphics/reference/DDSFileReference/ddsfileformat.htm
https://docs.microsoft.com/en-us/windows/win32/direct3ddds/dds-header

Perhaps you should contact the developer of the gimp-dds plugin and see if he has an update for your version of GIMP.

**Comment by IgeNiaI** [comment, marked as answer]
Thanks for help guys and sorry for not responding for so long. I think the answer is that IM cannot do what I need so we can close this ticket.
